File: applications/java/wordcount/WcReducer.java
package pmr;
import java.io.*;
import java.util.*;
import pmr.Reducer;

public class WcReducer {
	
	public static String join(String[] words, String delimiter )
	{
    	StringBuilder sb = new StringBuilder();
    	int i=0;
    	for (;i<words.length-1;i++)
    	{
    		sb.append(words[i] + ",");		        		
    	}
    	
    	sb.append(words[i]);
    	
    	return sb.toString();
	}

	public static void main(String[] args) throws Exception{

		    // Initialize Reduce job
		    Reducer reduceJob = new Reducer(args);         
		    
		    // Reduce function    
		    Map<String, Integer> count = new HashMap<String, Integer>();
		    
		    // split the map emitted to get words count from each partition file
		    for (String pName: reduceJob.partitionFiles)
		    {		    
		        BufferedReader inFile = new BufferedReader(new FileReader(pName));
		        String line;
		        String[] tokens;
		        while( (line = inFile.readLine()) != null )
		        {
		        	// Each line is delimited by comma by MR framework.
		        	tokens = line.split(",");
		        	
		        	// Actual word might contain "," and count is last token.
		        	int value = Integer.parseInt(tokens[tokens.length-1]);
		        	String word = join(Arrays.copyOfRange(tokens, 0, tokens.length-1), ",");
		        	
		        	if (count.containsKey(word))
		        		count.put(word, count.get(word) + value);
		        	else
		        		count.put(word, value);
		        }
		        inFile.close();
		    }
		    for(Map.Entry<String, Integer> e: count.entrySet())
		    	reduceJob.emit(e.getKey(),e.getValue());

		    // Finalize reduce job   
		    reduceJob.finalize(); 
		    
	}
}

File: applications/python/wordcount/wc_reducer.py
from pmr.reducer import Reducer
import sys


if __name__ == "__main__":
    # Initialize Reduce job
    reduceJob = Reducer(sys.argv)         
    
    # Reduce function    
    count = {}        
    # split the map emitted to get words count from each partition file                       
    for pName in reduceJob.partitionFiles:
        with open(pName) as infile:
            for line in infile:
                tokens = line.split(",")
                
                # Actual word might contain "," and count is last token. 
                value = int(tokens[-1])
                word = ",".join(tokens[:-1])
                
                if count.has_key(word):
                    count[word] = count[word] + value
                else:
                    count[word] = value

    for word, count in count.iteritems():                
        reduceJob.emit(word, count)                

    # Finalize reduce job   
    reduceJob.finalize() 
